Project 9

Meidoragon opened the pull request:
Update with changes specified by Sprint 9's project.

Meidoragon committed:
fix indentation

Meidoragon committed:
add card deletion confirmation modal

Meidoragon committed:
add avatar update modal

Meidoragon committed:
remove old debug log

Meidoragon committed:
add api constants

Meidoragon committed:
write API skeleton

Meidoragon committed:
add response handler to api class

Meidoragon committed:
add getCardList to API.js

Meidoragon committed:
add handling of empty card lists

Meidoragon committed:
add basic functionality for user added cards

Meidoragon committed:
modify names of variables to align with api

Meidoragon committed:
fix bug plaguing initial card rendering

Meidoragon committed:
change more variable names to fit api

Meidoragon committed:
update variable names

Meidoragon committed:
more bugfix for inital card load

Meidoragon committed:
add user acquiring user info to promises

Meidoragon committed:
set name and description from api

Meidoragon committed:
feat: updating profile changes serverside info

Meidoragon committed:
feat: proper card addition

it now sends that information to the server to be read later!

Meidoragon committed:
add asset for avatar edit button

Meidoragon committed:
fix: message displayed in confirmation popup

change from '' to 'Are you sure?'

Meidoragon committed:
feat: add confirmation popup outline

Meidoragon committed:
feat: style confirmation popup

including media queries

Meidoragon committed:
change: delete button leads to confirmation Popup

broke it so that it no longer deletes at all
anymore. to be fixed

Meidoragon committed:
fix: favicon.ico resulting in 404

Meidoragon committed:
feat: add ability to delete cards

affects both server and client

Meidoragon committed:
feat: add/remove 'likes'

Meidoragon committed:
feat: add updateAvatar to API.js

Meidoragon committed:
feat: add avatar hover state

Meidoragon committed:
feat: add avatar edit form

Meidoragon committed:
feat: add updating serverside avatar

Meidoragon committed:
feat: add avatar updates to show clientside

Meidoragon committed:
feat: add first ux improvement

avatar edit popup displays 'Saving...' while waiting for api response

Meidoragon committed:
feat: add part 2 of improving UX

the popup for profile editing now shows that
it is saving the information

Meidoragon committed:
feat: ux improvement part 3: card submission

Meidoragon committed:
feat: implement final ux improvement

Meidoragon committed:
clean and document code

kseniya57 commented on src/components/API.js:
```diff
@@ -0,0 +1,132 @@
+    this.#headers = options.headers; 
+  }
+
+  #handleResponse(response){
```
Great that you've implemented a  method for processing server response, so the code that checks the server response and converts it from JSON  is reused

kseniya57 commented on src/components/Card.js:
```diff
@@ -55,22 +70,36 @@ export default class Card{
+   * gets id generated by api
+   * @returns id
+   */
+  getCardId(){
```
It's great to have a getter for the card id

kseniya57 commented on src/components/Card.js:
```diff
@@ -1,41 +1,56 @@
 
   #handleLikeClick = () => {
+    this.#sendLikeInfo(this.#isLiked);
     this.#likeButton.classList.toggle('card__button_type_like-active');
```
Needs correction: You should toggle the like button only after the successful api response.

The code for rendering likes should look like this:

```js
// in Card class 

// call this method from index.js after successful api response in `then` block
setIsLiked(isLiked) { 
   // set instance variable
    this._isLiked = isLiked; 
    this._renderLikes(); 
} 

isLiked() {
  return this._isLiked;
}

_renderLikes() {
  if (this._isLiked) {
    // add active class to the like button
  } else {
    // remove active class from the like button
  }
}

createCard() { 
  ...
  // render initial likes, you saved them in the constructor
  this._renderLikes()
}

// index.js
if (card.isLiked()) {
  api.dislikeCard(...).then(response => card.setIsLiked(response._isLiked)).catch(() => /*handle error*/)
} else {
 // the same for adding like
}
```

kseniya57 commented on src/components/PopupConfirmation.js:
```diff
@@ -0,0 +1,42 @@
+   * @param {Object} card 
+   */
+  open(card){
+    this.#card = card;
```
Needs correction: `PopupWithConfirmation` is a general class. It can be used for any confirmation, not only card delete confirmation. This class should have a setter for `_handleFormSubmit`, not for card, so you can use a card in that handler in `index.js`.

```js
constructor(popupSelector) {
  super(popupSelector);
  this.#buttonElement = ..
}

setAction(action) { 
  this.#handleConfirm = action; 
} 

..
this.#buttonElement.addEventListener('mousedown', this.#handleConfirm)
```

kseniya57 commented on src/components/Section.js:
```diff
@@ -15,7 +15,7 @@ export default class Section {
    */
   constructor(items, renderer, element) {
     this.#renderedItems = items;
```
Needs correction: The items are not already rendered, use `items` as the name here

kseniya57 commented on src/components/UserInfo.js:
```diff
@@ -1,38 +1,58 @@
+    this.#description = userInfo.about;
+    this.#avatar = userInfo.avatar;
+
+    //display name, description, and avatar in page
```
Needs correction: You should set the contents via `  setUserInfo` and `setAvatar` methods when initializing the user

kseniya57 commented on src/components/UserInfo.js:
```diff
@@ -1,38 +1,58 @@
+    this.#avatarElement = elements.avatarElement;
+
+    //store name, description, and avatar in object
+    this.#name = userInfo.name;
```
Needs correction: There is no need to save the contents in the class instance, you can find it via the dom methods

kseniya57 commented on src/components/UserInfo.js:
```diff
@@ -1,38 +1,58 @@
-    this.#name = this.#nameElement.textContent;
-    this.#description = this.#descriptionElement.textContent
+  getUserInfo(){
+    return {username: this.#name, description: this.#description};
```
Needs correction: You should return the element contents here, for example `username:  this.#nameElement.textContent`

kseniya57 commented on src/components/UserInfo.js:
```diff
@@ -1,38 +1,58 @@
-  getUserInfo(){
-    return {name: this.#name, description: this.#description};
+  setUserInfo(newInfo){
+    this.#name = newInfo.name;
```
Needs correction: There is no need to save the contents in the class instance, you can find it via the dom methods

kseniya57 commented on src/components/UserInfo.js:
```diff
@@ -1,38 +1,58 @@
-    this.#nameElement.textContent = this.#name;
-    this.#descriptionElement.textContent = this.#description;
+  updateAvatar(newLink){
+    this.#avatar = newLink;
```
Needs correction: There is no need to save the contents in the class instance, you can find it via the dom methods

kseniya57 commented on src/pages/index.js:
```diff
@@ -51,41 +66,140 @@ enableValidation(options);
   evt.preventDefault();
-  user.setUserInfo(profileFormPopup.getInputValues());
-  profileFormPopup.close();
+  buttons.profileSubmitButton.textContent = 'Saving...'
```
Needs correction: The code for rendering loading gets duplicated.

You can manage loading via `PopupWithForm`:

```js
constructor(popupSelector, { loadingButtonText, ... }) { 
  ...
  this._buttonText = this._submitButton.textContent;
  this._loadingButtonText = loadingButtonText;
}

showLoading() {
  this._submitButton.textContent = this._loadingButtonText
}

hideLoading() {
  this._submitButton.textContent = this._buttonText
}
```

```js
// index.js
profileEditPopup.showLoading()
...
.finally(() => {
 profileEditPopup.hideLoading()
})
```

kseniya57 commented on src/pages/index.js:
```diff
@@ -51,41 +66,140 @@ enableValidation(options);
+      api.catchErrors(response);
+    })
+    .finally(() => {
+      profileFormPopup.close();
```
Needs correction: Сlose modal in `then` block, otherwise the form will close even if an error occurred and you will need to reopen the form and fill the values again.

kseniya57 commented on src/pages/index.js:
```diff
@@ -51,41 +66,140 @@ enableValidation(options);
+  buttons.confirmationButton.textContent = 'Deleting...'
+  api.deleteCard(card.getCardId())
+    .then((response) => {
+      card.getCardElement().remove();
```
Needs correction: you should use Card class methods for removing cards

kseniya57 commented on src/pages/index.js:
```diff
@@ -51,41 +66,140 @@ enableValidation(options);
+
+//Initialize API interactions
+const api = new API(apiOptions)
+Promise.all([api.getCardList(), api.getUserInfo()])
```
You've used `Promise.all` to query all the initial data at once. Awesome!

Meidoragon committed:
fix: like button toggle only after api response

per https://github.com/Meidoragon/se_project_aroundtheus/pull/2#discussion_r1314106282

Meidoragon committed:
fix: clarify unclear variable name

https://github.com/Meidoragon/se_project_aroundtheus/pull/2#discussion_r1314106885

Meidoragon committed:
fix: optimize userinfo class based on pr feedback

Meidoragon committed:
fix: reduce repeated code around popup interaction

per https://github.com/Meidoragon/se_project_aroundtheus/pull/2#discussion_r1314108010

Meidoragon committed:
fix: apply further feedback

Meidoragon committed:
fix: make PopupConfirmation less specific

specifically it no longer specifically for cards, and hypothetically
it should be usable for anything now.

Meidoragon commented:
I have addressed all items marked as 'needs correction'.
